docs/design/pr-builder.rst

```diff
@@ -0,0 +1,157 @@
+
+Pull requests can be Treated as a Type of Temporary ``Version``.
+We might consider adding a Boolean Field or ``VERSION_TYPES`` to the ``Version`` model.
+
+- If we go with Boolean Field we can add something like ``is_pull_request`` Field.
+- If we go with ``VERSION_TYPES`` we can add something like ``pull_request`` alongside Tag and Branch.
+
+We also have to update the current ``Version`` model ``QuerySet`` to exclude the PR versions.
+We have to add ``QuerySet`` for PR versions also.
+
+Excluding PR Versions from Elasticsearch Indexing
+-------------------------------------------------
+
+We should exclude to PR Versions from being Indexed to Elasticsearch.
+We need to update the queryset to exclude PR Versions.
+
+Creating Versions for Pull Requests
+-----------------------------------
+
+If the Github webhook event is ``pull_request`` and action is ``opened``,
+this means a pull request was opened in the projects repository.
+We can create a ``Version`` from the Payload data and trigger a initial build for the version.
+A version will be created whenever RTD receives an event like this.
+
+Triggering Build for New Commits in a Pull Request
+--------------------------------------------------
+
+We might want to trigger a new build for the PR version if there is a new commit on the PR.
+If the Github webhook event is ``pull_request`` and action is ``synchronize``,
+this means a new commit was added to the pull request.
+
+Status Reporting to Github
+--------------------------
+
+We could send build status reports to Github. We could send if the build was Successful or Failed.
+We can also send the build URL. By this we could show if the build passed or failed on Github something like travis-ci does.
+
+As we already have the ``repo:status`` scope on our OAuth App,
+we can send the status report to Github using the Github Status API.
+
+Sending the status report would be something like this:
+
+.. http:post:: /repos/:owner/:repo/statuses/:sha
+
+.. code:: json
+
+   {
+       "state": "success",
+       "target_url": "<pr_build_url>",
+       "description": "The build succeeded!",
+       "context": "continuous-documentation/read-the-docs"
+   }
+
+Storing Pull Request Docs
+-------------------------
+
+We need to think about how and where to store data after a PR Version build is finished.
+We can store the data in a blob storage.
+
+Deleting a PR Version
+---------------------
+
+We can delete a PR version when:
+
+- A pull request is ``closed``. Github Webhook event (Action: ``closed``, Merged: ``False``)
+- A pull request is ``merged``. Github Webhook event (Action: ``closed``, Merged: ``True``)
+- A PR Version has reached its life time
+
+We might want to set a life time of a PR version in case:
+
+- We don't receive webhook event for a pull request being closed/merged
+- If a PR is stale (not merged for a long time).
+
+We need to delete the PR Version alongside the Build data from the blob storage.
```

Comment:
I think we can just plan to keep the PR build data forever, and deal with it in the future if we have too much. I don't think that will be too big of an issue though.

Reply:
Okay! removed this part from the design doc.